src/nodes.coffee

```diff
@@ -1314,6 +1314,115 @@ exports.LineComment = class LineComment extends Base
+        # Catch invalid CSX attributes: <div {a:"b", props} {props} "value" />
+        if not (attr instanceof Obj or attr instanceof IdentifierLiteral) or (attr instanceof Obj and not attr.generated and (attrProps.length > 1 or not (attrProps[0] instanceof Splat)))
+          obj.error """
+            Unexpected token. Allowed CSX attributes are: id="val", src={source}, {props...} or attribute.
+          """
+        obj.base.csx = yes if obj.base instanceof Obj
+        fragments.push @makeCode ' '
+        fragments.push obj.compileToFragments(o, LEVEL_PAREN)...
+    if @content
+      fragments.push @makeCode('>')
+      fragments.push @content.compileNode(o, LEVEL_LIST)...
+      fragments.push [@makeCode('</'), tag..., @makeCode('>')]...
+    else
+      fragments.push @makeCode(' />')
+    fragments
+
+  ast: ->
+    # The location data spanning the opening element < ... > is captured by
+    # the generated Arr which contains the element's attributes
+    @openingElementLocationData = locationDataToAst @attributes.base.locationData
+
+    tagName = @tagName.base
+    tagName.locationData = tagName.tagNameLocationData
+    if @content?
+      @closingElementLocationData = mergeAstLocationData(
+        locationDataToAst tagName.closingTagOpeningBracketLocationData
+        locationDataToAst tagName.closingTagClosingBracketLocationData
+      )
+
+    super()
+
+  astType: ->
+    'JSXElement'
+
+  astProperties: ->
+    openingElement = Object.assign {
+      type: 'JSXOpeningElement'
+      name: @tagName.unwrap().ast()
+      selfClosing: not @closingElementLocationData?
+      attributes: []
+    }, @openingElementLocationData
+
+    closingElement = null
+    if @closingElementLocationData?
+      closingElement = Object.assign {
+        type: 'JSXClosingElement'
+        name: Object.assign(
+          @tagName.unwrap().ast(),
+          locationDataToAst @tagName.base.closingTagNameLocationData
+        )
+      }, @closingElementLocationData
+      if closingElement.name.type is 'JSXMemberExpression'
+        rangeDiff = closingElement.range[0] - openingElement.range[0] + '/'.length
+        shiftAstLocationData = (node) ->
+          node.range = [
+            node.range[0] + rangeDiff
+            node.range[1] + rangeDiff
+          ]
+          node.start += rangeDiff
+          node.end += rangeDiff
+          node.loc.start = {
+            line: node.loc.start.line
+            column: node.loc.start.column + rangeDiff
+          }
+          node.loc.end = {
+            line: node.loc.end.line
+            column: node.loc.end.column + rangeDiff
+          }
+        currentExpr = closingElement.name
+        while currentExpr.type is 'JSXMemberExpression'
+          shiftAstLocationData currentExpr unless currentExpr is closingElement.name
+          shiftAstLocationData currentExpr.property
+          currentExpr = currentExpr.object
+        shiftAstLocationData currentExpr
```

Comment:
So all of this `openingElement` and `closingElement` code is because we don’t have node classes for those?

Reply:
Right, now with the introduction of `CSXElement` we can use the normal AST methods for the outermost JSX element AST node (`JSXElement`), but the immediate AST children `JSXOpeningElement`/`JSXClosingElement` still don't have corresponding node classes so their AST nodes have to be constructed manually

I don't see a way to introduce separate node classes for the opening/closing elements since there aren't corresponding structures in our pseudo-representation of JSX (everything between and including an opening and closing element, ie a whole `JSXElement`, is represented by a single `Call`)